pgeocode.py

```diff
@@ -311,9 +311,61 @@ def query_postal_code(self, codes):
             response = response.iloc[0]
         return response
 
-    def query_location(self, name):
-        """Get locations information from a community/minicipality name"""
-        pass
+    def query_location(self, name: str, top_k: int = 500, fuzzy_threshold: int = 80,
+                       col: str = "place_name") -> pd.DataFrame:
+        """Get location information from a place name
+
+        Parameters
+        ----------
+        name: str
+          string containing place names to search for
+        top_k: int
+          maximum number of results (rows in DataFrame) to return
+        fuzzy_threshold: int
+          threshold (lower bound) for fuzzy string search for finding the place name
+        col: str
+          which column in the internal data to search through
+
+        Returns
+        -------
+        df : pandas.DataFrame
+          a pandas.DataFrame with the relevant information for all matching place names (or empty if no match was found)
+        """
+        contains_matches = self._str_contains_search(name, col)
+        if len(contains_matches) > 0:
+            return contains_matches.iloc[:top_k]
+
+        fuzzy_matches = self._fuzzy_search(name, col, threshold=fuzzy_threshold)
+        if len(fuzzy_matches) > 0:
+            return fuzzy_matches.iloc[:top_k]
+
+        return pd.DataFrame(columns=self._data.columns)
+
+    def _str_contains_search(self, text: str, col: str) -> pd.DataFrame:
+        candidates = self._data[col]
+        text_len = len(text)
+        match_mask = candidates.str.lower().str.contains(text.lower())
+        if match_mask.sum() == 0:
+            return pd.DataFrame(columns=self._data.columns)
```

Comment:
I don't understand what the lines below do,

Shouldn't this be just,
```
return self._data[match_mask]
```
including for an empty mask? But maybe I'm missing something.

Reply:
The reason for the if statement is that the following three lines of code should only be executed if there are any rows containing the queried place name. The following three lines ensure that the match with the closest string length to the input query string is returned (this only makes sense if there are any matches at all). The reasoning was that an end user may want to best string match, not all rows containing the string. But we can change it of course, then the end user is responsible for any further filtering in case there are multiple rows in the return dataframe.

Reply:
Adapted this as suggested